llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -5555,6 +5565,13 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
   return computeKnownBits(Op, Depth).isNonZero();
 }
 
+bool SelectionDAG::isKnownNonNegativeFP(SDValue Op) const {
+  if (ConstantFPSDNode *C1 = isConstOrConstSplatFP(Op, true))
+    return !C1->isNegative();
```

Comment:
Maybe need to guard against nans

Reply:
Since nans are allowed I would suggest being explicit and renaming to `isKnownPositiveOrNaNFP`